audio/out/buffer.c

```diff
@@ -117,14 +126,41 @@ struct mp_async_queue *ao_get_queue(struct ao *ao)
     return p->queue;
 }
 
+// called locked
+static void prune_queued_sameples(struct ao *ao)
+{
+    struct buffer_state *p = ao->buffer_state;
+    int pos;
+    double cur_time = MP_TIME_NS_TO_S(mp_time_ns());
+
+    for (pos = 0; pos <= p->num_queued_aframes; pos++) {
+        if (pos == p->num_queued_aframes || p->queued_aframes[pos].start_time > cur_time)
+            break;
+    }
+
+    // queued_aframes[pos - 1].start_time <= cur_time; keep it
+    // Remove 0 to pos - 2
+    if (pos > 1) {
+        int to_delete = pos - 1;
+        memmove(p->queued_aframes, p->queued_aframes + to_delete, sizeof(struct queued_aframe) * to_delete);
+        p->num_queued_aframes -= to_delete;
+    }
+}
+
 // Special behavior with data==NULL: caller uses p->pending.
-static int read_buffer(struct ao *ao, void **data, int samples, bool *eof,
+static int read_buffer(struct ao *ao, void **data, int samples, int64_t start_time_ns, bool *eof,
                        bool pad_silence)
 {
     struct buffer_state *p = ao->buffer_state;
     int pos = 0;
+
+    if (eof == NULL) {
+        eof = &(bool){0};
+    }
```

Comment:
the lifetime of eof ends here. It's not valid to use it beyond this point.

Reply:
i thought such constructs magically get converted to function-scoped.
or is that only for function args in calls?

Reply:
It has lifetime to enclosed block, which in this cases is this if.

Reply:
You are right. Since C99.